src/nomad_simulations/outputs.py

```diff
@@ -17,18 +17,48 @@
 #
 
 import numpy as np
-from typing import Optional
 from structlog.stdlib import BoundLogger
+from typing import Optional
 
 from nomad.datamodel.data import ArchiveSection
+from nomad.metainfo import Quantity, SubSection, MEnum, Section, Context
 from nomad.datamodel.metainfo.annotations import ELNAnnotation
-from nomad.metainfo import Quantity, SubSection, SectionProxy, Reference
 
-from .atoms_state import AtomsState, OrbitalsState
 from .model_system import ModelSystem
+from .physical_property import PhysicalProperty
 from .numerical_settings import SelfConsistency
 
 
+class ElectronicBandGap(PhysicalProperty):
+    """ """
+
+    rank = []
+
+    type = Quantity(
+        type=MEnum('direct', 'indirect'),
+        description="""
+        Type categorization of the electronic band gap. The electronic band gap can be `'direct'` or `'indirect'`.
+        """,
+    )
+
+    value = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        The value of the electronic band gap.
+        """,
+    )
+
+    # TODO add more functionalities here
+
+    def __init__(self, m_def: Section = None, m_context: Context = None, **kwargs):
+        super().__init__(m_def, m_context, **kwargs)
+        self.name = self.m_def.name
+
+    def normalize(self, archive, logger) -> None:
+        super().normalize(archive, logger)
+
+
```

Comment:
I guess this is a good example of what defining a new property looks like when you use `PhysicalProperty` as a base class. There are things here that need some thinking:

- We don't currently have a concept for what you are doing here with `rank`: section definitions only consist of fields for quantities and subsections, whereas `rank` here is some kind of a default value for all instances of this section. This may cause all kinds of problems in the current code and in the YAML schemas.
- Are the functions needed here? Remember also that if someone uses `PhysicalProperty` as a base section in YAML schemas, they cannot define functions.
- `value` is also kind of an interesting case: you need to re-refine it for each subclass (because the units or data type change) and should get an error if you don't define it, as it is part of the interface that all PhysicalProperties should follow.

Reply:
> value is also kind of an interesting case: you need to re-refine it for each subclass (because the units or data type change) and should get an error if you don't define it, as it is part of the interface that all PhysicalProperties should follow.

very good point! hadn't considered this potential pitfall yet

Reply:
> - We don't currently have a concept for what you are doing here with rank: section definitions only consist of fields for quantities and subsections, whereas rank here is some kind of a default value for all instances of this section. This may cause all kinds of problems in the current code and in the YAML schemas.

Not sure if I understand this part. `rank` **must** be defined by default for any physical property defined in the schema. This is just the rank of the tensor that we are refering to, so I don't know why any YAML schema should touch on this. Or do you mean that, defining by default like in here is not possible in a YAML?

> - Are the functions needed here? Remember also that if someone uses PhysicalProperty as a base section in YAML schemas, they cannot define functions.

Yes! This is so important, we also want to normalize the physical properties to maintain interoperability. This is a drawback of using YAML instead of Python schemas imo, not of what we defined here.

> - value is also kind of an interesting case: you need to re-refine it for each subclass (because the units or data type change) and should get an error if you don't define it, as it is part of the interface that all PhysicalProperties should follow.

Yes, this would be a bit better if `value` could take Any generic type, shape, unit at the `PhysicalProperty` level as well. But yes, once someone defines a new inherited PhysicalProperty, they need to overwrite `value` with the specific type and units (and shape via `rank`).


_____

> We should discuss this in the meeting next Wednesday. You could present this code as an example of what defining a new PhysicalProperty looks like as this is ultimately what users would need to use a lot in custom schemas.

Sure, there are a bunch of issues in this repo I want to tackle, to fully show the capabilities of `PhysicalProperty`, and how to use it in a complex case. I will try to have it for next week.